File: utils/diag.py
import sys
import os.path

if len(sys.argv) == 1:  # display help
    helpmsg = "Basic data diagnostic tool\n"
    helpmsg += "Usage:\npython "
    helpmsg += sys.argv[0] + "/path/to/cso_dir tol"
    sys.exit(helpmsg)
else:
    csodir = sys.argv[1]
    if 3 <= len(sys.argv):
        tol = sys.argv[2]
    else:
        tol = 100 # fallback to default value

i = 0
gap = 0
missing = 0
while True:
    basename = "%08d.cso" % i
    filepath = os.path.join(csodir,basename)
    cond_fail = not os.path.exists(filepath) or not os.path.isfile(filepath)
    if cond_fail:
        missing += 1
        gap += 1
        if gap > tol:
            print "Exceeded tol value of " + str(tol) + ", I give up searching"
            print "Found " + str(missing - tol - 1) + " files missing" 
            break
        else: pass
    else:
        print basename
        if gap > 0:
            gap = 0
    i += 1

File: utils/diag.cpp
/*********************************************************************/
/* g++ -std=c++11 -o diag diag.cpp -lboost_filesystem -lboost_system */
/*********************************************************************/

#include <iostream>
#include <cstdlib>
#include <sstream>
#include <boost/filesystem.hpp>

namespace fs = boost::filesystem;

int main(int argc, char *argv[]) {
  short tol = 100;
  char *csodir;

  if(argc == 1) {
    std::cout << "Basic data diagnostic tool" << std::endl
	      << "Usage:" << std::endl
	      << argv[0] << " /path/to/cso_dir tol" << std::endl;
    exit(EXIT_FAILURE);
  } else {
    csodir = argv[1];
    if(2 < argc) {
      std::istringstream ss(argv[2]);
      ss >> tol;
    }
  }

  int i = 0;
  int gap = 0;
  int missing = 0;
  std::string s;
  std::string basename;
  fs::path dir(csodir);
  fs::path filepath;

  while(1) {
    s = std::to_string(i);
    basename = std::string(8 - s.length(), '0').append(s);
    basename = basename.append(".cso");
    filepath = dir / basename;
    if (!fs::exists(filepath) || !fs::is_regular_file(filepath)) {
      missing++;
      gap++;
      if(gap > tol) {
	std::cout << "Exceeded tol value of " << tol << ", I give up searching" << std::endl;
	std::cout << "Found " << missing - tol - 1 << " files missing" << std::endl;
	break;
      }
    } else {
      std::cout << basename << std::endl;
      if(gap > 0) {
	gap = 0;
      }
    } 
    i++;
  }
  return 0;
}
